source/common/conn_pool/conn_pool_base.cc

```diff
@@ -233,9 +239,16 @@ void ConnPoolImplBase::onStreamClosed(Envoy::ConnectionPool::ActiveClient& clien
     // Close out the draining client if we no longer have active streams.
     client.close();
   } else if (client.state() == ActiveClient::State::Busy && client.currentUnusedCapacity() > 0) {
-    transitionActiveClientState(client, ActiveClient::State::Ready);
-    if (!delay_attaching_stream) {
-      onUpstreamReady();
+    if (!client.hasHandshakeCompleted()) {
+      transitionActiveClientState(client, ActiveClient::State::ReadyForEarlyData);
```

Comment:
Presumably non-http3 (and non-early-data) clients will never reach this method unless the handshake is already complete, right? (Just making sure I understand the flow)

Reply:
Correct